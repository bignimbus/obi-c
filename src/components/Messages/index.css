.messages {
  height: 0;
  padding: 0;
  position: relative;
  transition: height 250ms linear,
    padding 250ms linear;
  width: 100%;
}

.messages--show {
  box-sizing: border-box;
  height: 150px;
  padding: 8px;
}

.messages__ul {
  height: 100%;
  margin: 0 auto;
  padding: 0;
  width: 100%;
}

.messages__li {
  display: none;
  height: 100%;
}

.messages__li:first-of-type {
  display: flex;
  justify-content: space-between;
}

.messages__content {
  flex: 1 1 auto;
  padding-right: 8px;
}

.messages__footer {
  flex: 0 1 auto;
  display: flex;
  align-items: flex-end;
}

.messages__heading {
  margin: 0 auto;
  padding: 0;
}

.messages__body {
  margin: 0 auto;
  padding: 0;
}

.messages__button-container {
  height: 56px;
  width: 80px;
}
