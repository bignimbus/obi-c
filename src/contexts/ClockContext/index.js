import React, {
  useState,
  useEffect,
  createContext,
} from 'react';
import PropTypes from 'prop-types';
import User from '../../lib/User';
import Tick from '../../lib/Tick';

const ClockContext = createContext();

export const ClockContextProvider = ({ children }) => {
  const [user] = useState(new User({}));
  const [tick, setTick] = useState(new Tick({
    user,
    incrementStart: new Date(),
  }));
  const [activeNotifications, setActiveNotifications] = useState(tick.getActiveNotifications());

  const incrementClock = () => {
    const interval = setInterval(() => {
      const { incrementEnd } = tick;
      const fifteenSecondsAfterLastTickInterval = new Date(incrementEnd.getTime() + 15 * 1000);
      const now = new Date();
      if (fifteenSecondsAfterLastTickInterval > now) return;
      const newTick = new Tick({ user, incrementStart: incrementEnd });
      setTick(newTick);
      setActiveNotifications(newTick.getActiveNotifications());
    }, 1000);
    return clearInterval.bind(null, interval);
  };

  useEffect(incrementClock);

  return (
    <ClockContext.Provider
      value={{
        activeNotifications,
        addEvent: user.addEvent.bind(user),
        removeEvent: user.removeEvent.bind(user),
      }}
    >
      { children }
    </ClockContext.Provider>
  );
};

ClockContextProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

export default ClockContext;
