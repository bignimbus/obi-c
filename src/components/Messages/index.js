import React, { useContext } from 'react';
import Text from '../Text';
import ClockContext from '../../contexts/ClockContext';

const Messages = () => {
  const { removeEvent, activeNotifications } = useContext(ClockContext);

  return (
    <div>
      <ul>
        {
          activeNotifications.map(
            ({ title, body, notifiable }, i) => (
              <li key={i}>
                <section>
                  <h2>
                    <Text size='md'>
                      { title }
                    </Text>
                  </h2>
                  <p>
                    <Text size='sm'>
                      { body }
                    </Text>
                  </p>
                </section>
                <footer>
                  <button
                    onClick={() => {
                      removeEvent(notifiable);
                    }}
                  >
                    OK
                  </button>
                </footer>
              </li>
          ))
        }
      </ul>
    </div>
  );
};

export default Messages;
