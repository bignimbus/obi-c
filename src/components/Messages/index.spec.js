import React from 'react';
import renderer from 'react-test-renderer';
import Messages from '.';
import ClockContext from '../../contexts/ClockContext';

const removeEvent = jest.fn();

const renderComponent = activeNotifications => renderer.create(
  <ClockContext.Provider
    value={{
      removeEvent,
      activeNotifications,
    }}
  >
    <Messages />
  </ClockContext.Provider>,
);

describe('Messages', () => {
  it('should be able to render with no active notifications', () => {
    const element = renderComponent([]);
    expect(element.toJSON()).toMatchSnapshot();
  });

  it('should be able to render with active notifications', () => {
    const activeNotifications = [
      { title: 'Foo', body: 'bar baz', notifiable: 'I am a fake event' },
    ];
    const element = renderComponent(activeNotifications);
    expect(element.toJSON()).toMatchSnapshot();
  });
});
