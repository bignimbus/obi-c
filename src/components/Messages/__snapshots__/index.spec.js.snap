// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Messages should be able to render with active notifications 1`] = `
<div>
  <ul>
    <li>
      <section>
        <h2>
          <span
            className="
      text
      text--md
      
      
    "
          >
            Foo
          </span>
        </h2>
        <p>
          <span
            className="
      text
      text--sm
      
      
    "
          >
            bar baz
          </span>
        </p>
      </section>
      <footer>
        <button
          onClick={[Function]}
        >
          OK
        </button>
      </footer>
    </li>
  </ul>
</div>
`;

exports[`Messages should be able to render with no active notifications 1`] = `
<div>
  <ul />
</div>
`;
